### PR TITLE
chore: bump cppcheck to 2.13 version

### PR DESCRIPTION
**Pull request by knst**
## Issue being fixed or feature implemented
https://github.com/dashpay/dash-issues/issues/67

See also https://github.com/dashpay/dash/pull/5880 for prior work.
There's too many false alarm warnings with new kubuntu 23.10 (cppcheck 2.11)

## What was done?
Bump cppcheck to version 2.13 and fix related warnings or suppressing them if they are false-alarm.

## How Has This Been Tested?
Run `test/lint/lint-cppcheck-dash.sh` and see no output.

## Breaking Changes
N/A

## Checklist:
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

File: test/lint/lint-cppcheck-dash.sh [hunk 1]
```diff
@@ -31,25 +31,21 @@ IGNORED_WARNINGS=(
...
 
 # The following can be useful to ignore when the catch all is used
-#    "Consider performing initialization in initialization list."
     "Consider using std::transform algorithm instead of a raw loop."
     "Consider using std::accumulate algorithm instead of a raw loop."
     "Consider using std::any_of algorithm instead of a raw loop."
     "Consider using std::copy_if algorithm instead of a raw loop."
-#    "Consider using std::count_if algorithm instead of a raw loop."
-#    "Consider using std::find_if algorithm instead of a raw loop."
-#    "Member variable '.*' is not initialized in the constructor."
```
Comment (UdjinM6) [inline test/lint/lint-cppcheck-dash.sh]: should keep these as a reminder

File: src/sync.h [hunk 2]
```diff
@@ -244,7 +244,7 @@ using DebugLock = UniqueLock<typename std::remove_reference<typename std::remove
 #define LEAVE_CRITICAL_SECTION(cs)                                          \
     {                                                                       \
         std::string lockname;                                               \
-        CheckLastCritical((void*)(&cs), lockname, #cs, __FILE__, __LINE__); \
+        CheckLastCritical(reinterpret_cast<void*>(&cs), lockname, #cs, __FILE__, __LINE__); \
```
Comment (UdjinM6) [inline src/sync.h:247]: this code is not a Dash specific one, changes here can cause merge conflicts potentially

File: contrib/containers/ci/Dockerfile [hunk 3]
```diff
@@ -116,8 +116,8 @@ RUN apt-get update && apt-get install $APT_ARGS \
     wine64 \
     xorriso
 
-ARG CPPCHECK_VERSION=2.10
-RUN curl -sL "https://github.com/danmar/cppcheck/archive/${CPPCHECK_VERSION}.tar.gz" | tar -xvzf - --directory /tmp/
+ARG CPPCHECK_VERSION=2.13.0
+RUN curl -sL "https://codeload.github.com/danmar/cppcheck/tar.gz/refs/tags/${CPPCHECK_VERSION}" | tar -xvzf - --directory /tmp/
```
Comment (UdjinM6) [inline contrib/containers/ci/Dockerfile:120]: why changing url here?

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

File: contrib/containers/ci/Dockerfile (hunk 3 above)
Comment (UdjinM6) [reply contrib/containers/ci/Dockerfile:120]: I have no issues with the old url, try
https://github.com/danmar/cppcheck/archive/2.11.tar.gz
https://github.com/danmar/cppcheck/archive/2.13.0.tar.gz

Comment (PastaPastaPasta) [reply contrib/containers/ci/Dockerfile:120]: why this url diff still here?

**Comment by UdjinM6** [comment]
83f1cca00f7d538f308f4b83e803adfe678a4dcd is nice but it looks unrelated, it should probably be extracted into its own PR

**Comment by knst** [comment]
Otherwise 'cppcheck' complains: `CBlockPolicyEstimator& fee_estimator` can be `const`

**Review by UdjinM6 — DISMISSED** [review]
utACK

**Comment by knst** [comment]
> @[knst](https://github.com/knst) knst [force-pushed](https://github.com/dashpay/dash/compare/47c1b38c3296f22f6468e21fdc380d4d6518d58c..91b40e8683b8901c5acb42f672710c8939fb3781) the fix-cppcheck-2.13 branch from[ 47c1b38 ](https://github.com/dashpay/dash/commit/47c1b38c3296f22f6468e21fdc380d4d6518d58c)to 91b40e8

rebased to be on top of re-merged master.

**Comment by knst** [comment]
> @[knst](https://github.com/knst) knst [force-pushed](https://github.com/dashpay/dash/compare/91b40e8683b8901c5acb42f672710c8939fb3781..da54afb5b571a2aa3d0e81c7c622f89878a5fb93) the fix-cppcheck-2.13 branch from[ 91b40e8 ](https://github.com/dashpay/dash/commit/91b40e8683b8901c5acb42f672710c8939fb3781)to da54afb

lost one change accidentally, fixed.

**Comment by PastaPastaPasta** [comment]
Rebase looks good
```
 1:  6d80a3a010 !  1:  3f1cb8d13d test: add suppression for state.Invalid() which is widely used by cppcheck is unhappy
    @@ test/lint/lint-cppcheck-dash.sh: IGNORED_WARNINGS=(
      )
      
      # We should attempt to update this with all dash specific code
    -@@ test/lint/lint-cppcheck-dash.sh: FILES=$(git ls-files -- "src/batchedlogger.*" \
    -                         "src/evo/*.h" \
    -                         "src/governance/*.cpp" \
    -                         "src/governance/*.h" \
    --                        "src/hdchain.*" \
    -                         "src/keepass.*" \
    -                         "src/llmq/*.cpp" \
    -                         "src/llmq/*.h" \
     @@ test/lint/lint-cppcheck-dash.sh: FILES=$(git ls-files -- "src/batchedlogger.*" \
                              "src/rpc/governance.cpp" \
                              "src/rpc/masternode.cpp" \
    @@ test/lint/lint-cppcheck-dash.sh: FILES=$(git ls-files -- "src/batchedlogger.*" \
                              "src/statsd_client.*" \
                              "src/test/block_reward_reallocation_tests.cpp" \
     @@ test/lint/lint-cppcheck-dash.sh: FILES=$(git ls-files -- "src/batchedlogger.*" \
    -                         "src/test/dynamic_activation*.cpp" \
                              "src/test/evo*.cpp" \
                              "src/test/governance*.cpp" \
    +                         "src/wallet/hdchain.*" \
     -                        "src/unordered_lru_cache.h")
     +                        "src/unordered_lru_cache.h" \
    -+                        "src/wallet/hdchain.*" \
     +                        )
      
      
 2:  9f15182aaa =  2:  7daa9bea9a refactor: use std::any_of
 3:  f76110e716 =  3:  c23c25d00b test: supress any_of suggestions in cppcheck-dash
 4:  5d6f25fcba =  4:  f36bb1f085 style: add semicolumn that the end to ENTER/LEAVE CRITICAL SECTION
 5:  58dedf8329 =  5:  3abeab16d3 refactor: replace multiple C-style casts to reinterpret_cast
 6:  7f9484226a =  6:  28f7ecc11f test: tidy up and reorder warnings in lint-cppcheck-dash
 7:  44f40a71ec =  7:  61a6c407fe fix: re-order asserts in creditpool
 8:  aea366c8c8 =  8:  add99ea862 refactor: add more consts everywhere as required by cppcheck 2.13.0
 9:  731256e136 =  9:  b4ed65e15e refactor: add multiple missing `const`
10:  02f654676a = 10:  6bc14a35e5 chore: bump cpp check 2.10 to 2.13
11:  2ee5719017 = 11:  e7e355ba8b refactor: make SetNull in CJ classes virtual to prevent warning from compiler
12:  727fba31d2 = 12:  e2ac86b8f7 refactor: drop dependency of CJ to fee_estimator
13:  da54afb5b5 = 13:  df554f396b test: add multiple suppression for cppcheck to make it finally quiet
~
```

File: test/lint/lint-cppcheck-dash.sh [hunk 4]
```diff
@@ -88,7 +99,8 @@ FILES=$(git ls-files -- "src/batchedlogger.*" \
                         "src/test/evo*.cpp" \
                         "src/test/governance*.cpp" \
                         "src/wallet/hdchain.*" \
-                        "src/unordered_lru_cache.h")
+                        "src/unordered_lru_cache.h" \
+                        )
```
Comment (UdjinM6) [inline test/lint/lint-cppcheck-dash.sh:103]: could also reorder them while at it i.e.
```
                        "src/unordered_lru_cache.h" \
                        "src/wallet/hdchain.*" \
                        )
```

**Review by UdjinM6 — APPROVED** [review]
utACK

**Review by PastaPastaPasta — APPROVED** [review]
utACK

**Pull request merged**